[TP] Add prepareInput and output for input/output DTensor layout annotation in the parent module in TP API

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111176
* __->__ #111166
* #111160

In some use cases, we found that users might want to annote the input/output DTensor layout for the parent module rather than the submodule whose parameters are to be distributed so that we want to have these two class for users to annote input/output DTensor layouts so that we register pre-FWD/FWD hook for the TP-lized module.

fduwjj committed:
Add prepareInput and output for hook register in TP API

[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -565,3 +569,40 @@ def __init__(
+
+class PrepareModuleInput(ParallelStyle):
+    """
+    Only used to specify the input distribute spec for a module.
```
we should make this documentation clearer, we should mention sth like:
PrepareModuleInput allows user to annotate the input torch.Tensor layouts to create DTensors base on input_layouts, and prepare the input DTensors be transformed to a desired layout base on output layouts. 

We should also explain each kwarg separately, and possibly add a small example demonstrate: annotate input torch.Tensor as a Sharded DTensor, then transform it to a Replicated DTensor

fduwjj replied:
Added comments, args doc and examples.

fduwjj committed:
Update on "Add prepareInput and output for hook register in TP API"




[ghstack-poisoned]

fduwjj committed:
Update on "Add prepareInput and output for hook register in TP API"




[ghstack-poisoned]

fduwjj committed:
Update on "Add prepareInput and output for hook register in TP API"




[ghstack-poisoned]

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -589,3 +591,102 @@ def __init__(
+            input_layouts=input_layouts,
+            output_layouts=output_layouts,
+            use_local=use_local,
+            _prepare_input=_get_prepare_input(
```
a bit surprised it works as `_prepare_input/output` are positional args and it should appear in the front..

fduwjj replied:
but positional args can also be passed in as kwargs I guess?

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -589,3 +591,102 @@ def __init__(
+        >>> from torch.distributed.tensor.parallel import parallelize_module, PrepareModuleInput
+        >>> ...
+        >>> parallelize_plan = {
+        >>>     "attn": PrepareModuleInput(),   # The input of attn will be converted to Sharded DTensor.
```
nit: input of attn will be converted to Sharded DTensor and redistributed to Replicated DTensor?

wanchaol commented on torch/distributed/tensor/parallel/style.py:
```diff
@@ -589,3 +591,102 @@ def __init__(
+    >>> from torch.distributed.tensor.parallel import parallelize_module, PrepareModuleOutput
+    >>> ...
+    >>> parallelize_plan = {
+    >>>     "submodule": PrepareModuleOutput(),   # The input of attn will be converted to Sharded tensor.
```
nit: the output of attn will be converted to Replicated DTensor if it's not a DTensor, then redistributed to Sharded local tensor

wanchaol reviewed (approved):
lgtm, small nits inlined

fduwjj committed:
Update on "[TP] Add prepareInput and output for input/output DTensor layout annotation in the parent module in TP API"


In some use cases, we found that users might want to annote the input/output DTensor layout for the parent module rather than the submodule whose parameters are to be distributed so that we want to have these two class for users to annote input/output DTensor layouts so that we register pre-FWD/FWD hook for the TP-lized module.



[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Add prepareInput and output for input/output DTensor layout annotation in the parent module in TP API"


In some use cases, we found that users might want to annote the input/output DTensor layout for the parent module rather than the submodule whose parameters are to be distributed so that we want to have these two class for users to annote input/output DTensor layouts so that we register pre-FWD/FWD hook for the TP-lized module.



[ghstack-poisoned]

fduwjj committed:
Update on "[TP] Add prepareInput and output for input/output DTensor layout annotation in the parent module in TP API"


In some use cases, we found that users might want to annote the input/output DTensor layout for the parent module rather than the submodule whose parameters are to be distributed so that we want to have these two class for users to annote input/output DTensor layouts so that we register pre-FWD/FWD hook for the TP-lized module.



[ghstack-poisoned]

fduwjj commented on docs/source/distributed.tensor.parallel.rst:
```diff
@@ -53,6 +53,9 @@ used for input/output preparation:
 .. autofunction::  make_output_shard_1d
 .. autofunction::  make_output_tensor
 
+.. autofunction::  PrepareModuleInput
```
Add it here to pass CI first, Will have more doc changes in the next PR https://github.com/pytorch/pytorch/pull/111176

fduwjj committed:
Update on "[TP] Add prepareInput and output for input/output DTensor layout annotation in the parent module in TP API"


In some use cases, we found that users might want to annote the input/output DTensor layout for the parent module rather than the submodule whose parameters are to be distributed so that we want to have these two class for users to annote input/output DTensor layouts so that we register pre-FWD/FWD hook for the TP-lized module.



[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
